File: nodegraph/p1_cpp_nodegraph/s1_static_graph.cpp
#include<memory>
#include<list>
#include<iostream>

// local namespace
namespace {

/// @brief General node interface
class Node
{
public:
    // virtual ~Node() {} -- instead we define private dtor
    //                    -- note that Process() is still a virtual method
    //                    -- look at Pin<T> and it calls Process() using vtable
    virtual void ProcessForwards() = 0;
    virtual void ProcessBackwards() = 0;

protected:
    ~Node() {}
};

/// @brief Universal pin
/// @tparam T Type of data stored in this pin
template<class T> class Pin
{
public:
    /// @brief Obtain owning node
    Node &GetOwningNode() { return m_node; }

    /// @brief Obtain connected pin
    Pin<T> *GetConnectedPin() { return m_connection; }

    /// @brief Set data stored in this pin
    template<class X> void SetData(X &&x) { m_data = std::forward<X>(x); }

    /// @brief Get data stored in this pin
    T const & GetData() const { return m_data; }

    /// @brief Disconnect other pin
    void Disconnect()
    {
        if (m_connection)
        {
            m_connection->m_connection = nullptr;
            m_connection = nullptr;
        }
    }

private:
    Node &m_node;
    Pin<T> *m_connection;
    T m_data;

protected:
    /// @brief Create pin attached to specific node
    Pin(Node &node): m_node(node), m_connection(nullptr), m_data{} {}

    /// @brief Connect pin to another pin
    void Connect(Pin<T> &pin)
    {
        m_connection = &pin;
        pin.m_connection = this;
    }
};

// Forward declaration of OutputPin
template<class T> struct OutputPin;

/// @brief Input pin
/// @tparam T Type of data input
template<class T> struct InputPin : Pin<T>
{
    InputPin(Node &node): Pin<T>(node) {}

    /// @brief Connect output pin (implemented below to solve catch22)
    void Connect(OutputPin<T> &pin);

    /// @brief Receive data from the other connected pin into this pin
    void ReceiveData()
    {
        if (auto connectedPin = Pin<T>::GetConnectedPin(); nullptr != connectedPin)
        {
            Pin<T>::SetData(connectedPin->GetData());
        }
    }
    
    /// @brief Propagate processing to connected node
    void PropagateBackwards()
    {
        if (auto connectedPin = Pin<T>::GetConnectedPin(); nullptr != connectedPin)
        {
            connectedPin->GetOwningNode().ProcessBackwards();
        }
    }
};

/// @brief Output pin
/// @tparam T Type of data output
template<class T> struct OutputPin : Pin<T>
{
    OutputPin(Node &node): Pin<T>(node) {}

    /// @brief Connect output pin
    void Connect(InputPin<T> &pin) { Pin<T>::Connect(pin); }

    /// @brief Send data from this pin to the other connected pin
    void SendData()
    {
        if (auto connectedPin = Pin<T>::GetConnectedPin(); nullptr != connectedPin)
        {
            connectedPin->SetData(Pin<T>::GetData());
        }
    }

    /// @brief Propagate processing to connected node
    void PropagateForwards()
    {
        if (auto connectedPin = Pin<T>::GetConnectedPin(); nullptr != connectedPin)
        {
            connectedPin->GetOwningNode().ProcessForwards();
        }
    }
};

// Implementation of InputPin::Connect
template<class T> void InputPin<T>::Connect(OutputPin<T> &pin) { Pin<T>::Connect(pin); }

/// @brief Node that will provide data
/// @tparam T Type of source data
template<class T> class SourceNode : Node
{
public:
    ~SourceNode() {}
    SourceNode() : m_outputPin{*this}
    {}

    void ProcessForwards() override
    {
        GetOutputPin().SendData();
        GetOutputPin().PropagateForwards();
    }
    
    void ProcessBackwards() override
    {
        // This could be loading new value from somewhere
    }

    template<class X> void LoadValue(X &&x)
    {
        GetOutputPin().SetData(std::forward<X>(x));
    }

    OutputPin<T> &GetOutputPin() { return m_outputPin; }

private:
    OutputPin<T> m_outputPin;
};

/// @brief Node that will receive final result of all procesing
/// @tparam T Type of final result data
template<class T> class TargetNode : Node
{
public:
    ~TargetNode() {}
    TargetNode() : m_inputPin{*this}
    {}

    void ProcessForwards() override
    {
        // This could render result somewhere
    }

    void ProcessBackwards() override
    {
        GetInputPin().PropagateBackwards();
        GetInputPin().ReceiveData();
    }

    const T &GetValue() const 
    {
        // Process(); -- should we pull?
        return GetInputPin().GetData();
    }

    InputPin<T> &GetInputPin() { return m_inputPin; }
    InputPin<T> const &GetInputPin() const { return m_inputPin; }

private:
    InputPin<T> m_inputPin;
};

template<class X, class Y, class F> class TransformNode : Node
{
public:
    ~TransformNode() {}
    TransformNode(F f): m_function(std::move(f)), m_inputPin{*this}, m_outputPin{*this} {}
    
    void ProcessBackwards() override
    {
        // If we're propagating backwards, then request came from output pin
        // This can happen when we're pulling from output
        GetInputPin().PropagateBackwards();
        GetInputPin().ReceiveData();
        GetOutputPin().SetData(m_function(GetInputPin().GetData()));
    }

    void ProcessForwards() override
    {
        // If we're propagate forwards, then request came from input pin
        // This can happen when we're pushing new input data
        GetOutputPin().SetData(m_function(GetInputPin().GetData()));
        GetOutputPin().SendData();
        GetOutputPin().PropagateForwards();
    }

    InputPin<X> &GetInputPin() { return m_inputPin; }
    OutputPin<Y> &GetOutputPin() { return m_outputPin; }

private:
    F m_function;
    InputPin<X> m_inputPin;
    OutputPin<Y> m_outputPin;
};

} // end of local namespace

void test_int_to_double()
{
    // Let's declare some transformation function: int -> double
    auto func = [](int x) { return x * 1.5; };

    // Let's create node using our transformation function
    TransformNode<int, double, decltype(func)> transformNode{func};
    
    // Now we need some source
    SourceNode<int> sourceNode;

    // And we also need some target
    TargetNode<double> targetNode;

    transformNode.GetInputPin().Connect(sourceNode.GetOutputPin());
    transformNode.GetOutputPin().Connect(targetNode.GetInputPin());

    // Here we will push the value from the source
    sourceNode.LoadValue(10.0);
    sourceNode.ProcessForwards();
    
    std::cout << "Result of processing forwards: f(" << sourceNode.GetOutputPin().GetData() << ") => " << targetNode.GetValue() << std::endl;
    
    // Here we will pull the value from the target
    sourceNode.LoadValue(50.0);
    targetNode.ProcessBackwards();
    
    std::cout << "Result of processing backwards: f(" << sourceNode.GetOutputPin().GetData() << ") => " << targetNode.GetValue() << std::endl;

    // We prevent by design storage of the node that would causing type erasure
    //Node *ptr = &transformNode; -- This is invalid as TransformNode inherits Node privately
    //delete ptr; -- This is invalid as ~Node is private
}

template<class T>
std::ostream &operator <<(std::ostream &os, std::tuple<T, T> const &x)
{
    const auto [a, b] = x;
    return os << "[" << a << ", " << b << "]";
}

void test_with_shared_ptr()
{
    auto func = [](std::shared_ptr<std::tuple<int, int>> const &x)
    {
        const auto [a, b] = *x;
        return std::make_shared<std::tuple<double, double>>(a * b, a / (double)b);
    };

    // Let's create node using our transformation function
    TransformNode<
        std::shared_ptr<std::tuple<int, int>>,
        std::shared_ptr<std::tuple<double, double>>,
        decltype(func)> transformNode{func};
    
    // Now we need some source
    SourceNode<std::shared_ptr<std::tuple<int, int>>> sourceNode;

    // And we also need some target
    TargetNode<std::shared_ptr<std::tuple<double, double>>> targetNode;

    transformNode.GetInputPin().Connect(sourceNode.GetOutputPin());
    transformNode.GetOutputPin().Connect(targetNode.GetInputPin());

    // Here we will push the value from the source
    sourceNode.LoadValue(std::make_shared<std::tuple<int, int>>(7, 9));
    sourceNode.ProcessForwards();
    
    std::cout << "Result of processing forwards: f(" << *sourceNode.GetOutputPin().GetData() << ") => " << *targetNode.GetValue() << std::endl;
    
    // Here we will pull the value from the target
    sourceNode.LoadValue(std::make_shared<std::tuple<int, int>>(11, 5));
    targetNode.ProcessBackwards();
    
    std::cout << "Result of processing backwards: f(" << *sourceNode.GetOutputPin().GetData() << ") => " << *targetNode.GetValue() << std::endl;
}

File: nodegraph/p2_rust_nodegraph/src/s1_static_graph.rs
use std::cell::RefCell;

pub trait GraphNode<'a>
{
    fn process_forwards(&'a self);
    fn process_backwards(&'a self);
}

#[derive(Default)]
struct GraphNodePin<'a, T> where T: Clone + Default
{
    node: RefCell<Option<&'a dyn GraphNode<'a>>>,
    connection: RefCell<Option<&'a GraphNodePin<'a, T>>>,
    data: RefCell<T>
}

impl<'a, T> GraphNodePin<'a, T> where T: Clone + Default
{
    pub fn get_node(&'a self) -> Option<&'a dyn GraphNode<'a>> {
        *self.node.borrow()
    }

    pub fn get_connection(&'a self) -> Option<&'a GraphNodePin<'a, T>> {
        *self.connection.borrow()
    }

    pub fn setup(&'a self, node: &'a dyn GraphNode<'a>) {
        self.node.replace(node.into());
    }

    pub fn connect(&'a self, pin: &'a GraphNodePin<'a, T>) {
        self.connection.replace(pin.into());
    }

    pub fn set_data(&'a self, data: T) {
        self.data.replace(data);
    }

    pub fn get_data(&'a self) -> T {
        self.data.borrow().clone()
    }
}

#[derive(Default)]
pub struct GraphNodeInputPin<'a, T> where T: Clone + Default
{
    pin: GraphNodePin<'a, T>
}

impl<'a, T> GraphNodeInputPin<'a, T> where T: Clone + Default
{
    pub fn connect(&'a self, output_pin: &'a GraphNodeOutputPin<'a, T>) {
        self.pin.connect(&output_pin.pin);
        output_pin.pin.connect(&self.pin);
    }

    pub fn receive_data(&'a self) {
        if let Some(input) = self.pin.get_connection() {
            self.pin.set_data(input.get_data());
        }
    }

    pub fn propagate_backwards(&'a self) {
        if let Some(input) = self.pin.get_connection() {
            if let Some(node) = input.get_node() {
                node.process_backwards();
            }
        }
    }
}

#[derive(Default)]
pub struct GraphNodeOutputPin<'a, T> where T: Clone + Default
{
    pin: GraphNodePin<'a, T>
}

impl<'a, T> GraphNodeOutputPin<'a, T> where T: Clone + Default
{
    pub fn connect(&'a self, input_pin: &'a GraphNodeInputPin<'a, T>) {
        self.pin.connect(&input_pin.pin);
        input_pin.pin.connect(&self.pin);
    }

    pub fn send_data(&'a self) {
        if let Some(output) = self.pin.get_connection() {
            output.set_data(self.pin.get_data());
        }
    }
    
    pub fn propagate_forwards(&'a self) {
        if let Some(output) = self.pin.get_connection() {
            if let Some(node) = output.get_node() {
                node.process_forwards();
            }
        }
    }
}

#[derive(Default)]
pub struct SourceGraphNode<'a, T> where T: Clone + Default
{
    pub output_pin: GraphNodeOutputPin<'a, T>
}

impl<'a, T> SourceGraphNode<'a, T> where T: Clone + Default
{
    pub fn setup(&'a self) {
        self.output_pin.pin.setup(self);
    }

    pub fn load_value(&'a self, data: T)
    {
        self.output_pin.pin.set_data(data);
    }
}

impl<'a, T> GraphNode<'a> for SourceGraphNode<'a, T> where T: Clone + Default
{
    fn process_backwards(&'a self) {
    }

    fn process_forwards(&'a self) {
        self.output_pin.send_data();
        self.output_pin.propagate_forwards();
    }
}

#[derive(Default)]
pub struct TargetGraphNode<'a, T> where T: Clone + Default
{
    pub input_pin: GraphNodeInputPin<'a, T>
}

impl<'a, T> TargetGraphNode<'a, T> where T: Clone + Default
{
    pub fn setup(&'a self) {
        self.input_pin.pin.setup(self);
    }

    pub fn get_value(&'a self) -> T
    {
        self.input_pin.pin.get_data()
    }
}

impl<'a, T> GraphNode<'a> for TargetGraphNode<'a, T> where T: Clone + Default
{
    fn process_backwards(&'a self) {
        self.input_pin.propagate_backwards();
        self.input_pin.receive_data();
    }

    fn process_forwards(&'a self) {
    }
}

pub struct TransformGraphNode<'a, X, Y, F>
where
    X: Clone + Default,
    Y: Clone + Default,
    F: Fn(X) -> Y
{
    pub input_pin: GraphNodeInputPin<'a, X>,
    pub output_pin: GraphNodeOutputPin<'a, Y>,
    pub func: F
}

impl<'a, X, Y, F> TransformGraphNode<'a, X, Y, F>
where
    X: Clone + Default,
    Y: Clone + Default,
    F: Fn(X) -> Y
{
    pub fn new(func: F) -> Self {
        Self {
            input_pin: Default::default(),
            output_pin: Default::default(),
            func
        }
    }
    
    pub fn setup(&'a self) {
        self.input_pin.pin.setup(self);
        self.output_pin.pin.setup(self);
    }
}

impl<'a, X, Y, F> GraphNode<'a> for TransformGraphNode<'a, X, Y, F>
where
    X: Clone + Default,
    Y: Clone + Default,
    F: Fn(X) -> Y
{
    fn process_backwards(&'a self) {
        self.input_pin.propagate_backwards();
        self.input_pin.receive_data();
        self.output_pin.pin.set_data((self.func)(self.input_pin.pin.get_data()));
    }

    fn process_forwards(&'a self) {
        self.output_pin.pin.set_data((self.func)(self.input_pin.pin.get_data()));
        self.output_pin.send_data();
        self.output_pin.propagate_forwards();
    }
}


#[test]
fn test_me() {
    let foo = |x: i32| -> i32 { x * x };

    let source_node = SourceGraphNode::default();
    let target_node = TargetGraphNode::default();
    let transform_node = TransformGraphNode::new(foo);

    source_node.setup();
    target_node.setup();
    transform_node.setup();

    transform_node.input_pin.connect(&source_node.output_pin);
    transform_node.output_pin.connect(&target_node.input_pin);

    source_node.load_value(10);
    source_node.process_forwards();
    let value1 = target_node.get_value();

    println!("Result of process forwards: {}", value1);
    
    source_node.load_value(20);
    target_node.process_backwards();
    let value2 = target_node.get_value();
    
    println!("Result of process backwards: {}", value2);
}